functions-python/gbfs_validator/src/gbfs_utils.py

```diff
@@ -0,0 +1,155 @@
+import json
+import logging
+import os
+import uuid
+from datetime import datetime
+from typing import Dict, Any
+
+import requests
+from google.cloud import storage
+
+from database_gen.sqlacodegen_models import (
+    Gbfssnapshot,
+    Gbfsvalidationreport,
+    Gbfsnotice,
+)
+from dataset_service.main import (
+    Status,
+)
+
+VALIDATOR_URL = os.getenv(
+    "VALIDATOR_URL",
+    "https://gbfs-validator.mobilitydata.org/.netlify/functions/validator-summary",
+)
+
+
+def fetch_gbfs_files(url: str) -> Dict[str, Any]:
+    """Fetch the GBFS files from the autodiscovery URL."""
+    response = requests.get(url)
+    response.raise_for_status()
+    return response.json()
+
+
+def upload_gbfs_file_to_bucket(
+    bucket: storage.Bucket, file_url: str, destination_blob_name: str
+) -> str:
+    """Upload a GBFS file to a Cloud Storage bucket."""
+    response = requests.get(file_url)
+    response.raise_for_status()
+    blob = bucket.blob(destination_blob_name)
+    blob.upload_from_string(response.content)
+    blob.make_public()
+    logging.info(f"Uploaded {destination_blob_name} to {bucket.name}.")
+    return blob.public_url
+
+
+def create_gbfs_json_with_bucket_paths(
+    bucket: storage.Bucket, gbfs_data: Dict[str, Any], stable_id: str
+) -> str:
+    """
+    Create a new gbfs.json with paths pointing to Cloud Storage and upload it.
+    @param bucket: The Cloud Storage bucket.
+    @param gbfs_data: The GBFS data.
+    @param stable_id: The stable ID of the feed.
+    @return: The public URL of the new gbfs.json.
+    """
+    new_gbfs_data = gbfs_data.copy()
+    today = datetime.now().strftime("%Y-%m-%d")
+
+    for feed_key, feed in new_gbfs_data["data"].items():
+        if isinstance(feed["feeds"], dict):
+            for feed_language, feed_info in feed["feeds"].items():
+                old_url = feed_info["url"]
+                blob_name = f"{stable_id}/{stable_id}-{today}/{feed_info['name']}_{feed_language}.json"
+                new_url = upload_gbfs_file_to_bucket(bucket, old_url, blob_name)
+                feed_info["url"] = new_url
+        elif isinstance(feed["feeds"], list):
+            for feed_info in feed["feeds"]:
+                old_url = feed_info["url"]
+                blob_name = f"{stable_id}/{stable_id}-{today}/{feed_info['name']}.json"
+                new_url = upload_gbfs_file_to_bucket(bucket, old_url, blob_name)
+                feed_info["url"] = new_url
+        else:
+            logging.warning(f"Unexpected format in feed: {feed_key}")
```

Comment:
[suggestion]: I would suggest adding a try/catch section to get a detailed error message if one of the URLs fails while downloading.